Manifest Based Image Authorization

mhatrevi opened the pull request:
This change contains the Manifest based Image Authorization feature.
Design document: https://docs.google.com/document/d/1BofLEAzeszgAGVcaEjTxM7tDxwIq9KhVXuqGUlrg1sc/edit#heading=h.z8ehugjyw5jp

mhatrevi committed:
Authorization Manifest: Initial Changes

mhatrevi committed:
AUTH_AND_STASH

mhatrevi committed:
Adding Image Metadata collection signature verification

mhatrevi committed:
Adding lock file

mhatrevi committed:
Merge branch 'main' into vmhatre/man-auth

mhatrevi committed:
Fixing clippy errors

nquarton commented on runtime/tests/runtime_integration_tests/test_set_auth_manifest.rs:
```diff
@@ -0,0 +1,128 @@
+use caliptra_runtime::RtBootStatus;
+use zerocopy::AsBytes;
+
+#[test]
```
Is there a plan to add a test for each of the possible errors above? (mismatched signatures, mismatched marker, size bounds, etc.)

mhatrevi committed:
Merge branch 'main' into vmhatre/man-auth

nquarton commented:
I think this also needs the commands documented in the runtime readme

ArthurHeymans committed:
Make the preamble non persistent (#1619)

This change does the following:
- Increase runtime region size. The ICCM is 128K so this still leave 11K
  for the FMC or the runtime to grow in the future
- Fix asserts on layout checks
- Re-add removed prints
- Use .get instead of array indexing to avoid the generation of panics

mhatrevi commented on runtime/src/authorize_and_stash.rs:
```diff
@@ -0,0 +1,103 @@
+
+            let mut auth_result = DENY_IMAGE_AUTHORIZATION;
+            for metadata_entry in auth_manifest_image_metadata_col.image_metadata_array.iter() {
+                if metadata_entry.digest == cmd.measurement {
```
Good point. We'll add CFI protections to this.

ArthurHeymans committed:
drivers/persistent: Ensure no rom changes with AuthManifest (#1628)

Adding the data structures to the code somehow changes the ROM so guard
those as those structs as those are only used in runtime anyway. 
A similar thing is done for the PCR reset counter.

nquarton commented on runtime/src/packet.rs:
```diff
@@ -25,7 +25,7 @@ pub struct Packet {
 }
 
-const MAX_PAYLOAD_SIZE: usize = 512; // in dwords
+const MAX_PAYLOAD_SIZE: usize = 2050; // in dwords
```
Okay cool. I didn't realize it was that large. Thanks!

ArthurHeymans commented on api/src/mailbox.rs:
```diff
@@ -918,6 +935,126 @@ impl Request for QuotePcrsReq {
+#[derive(Debug, Default, AsBytes, FromBytes, PartialEq, Eq)]
+pub struct AuthorizeAndStashResp {
+    pub hdr: MailboxRespHeader,
+    pub auth_req_result: u32,
```
It's not a DPE result.

nquarton replied:
Got it. In that case, should we be returning the DPE result as we do for stash measurement?

mhatrevi replied:
I don't think an additional DPE result is needed. @JohnTraverAmd for confirmation.

mhatrevi commented on auth-manifest/app/src/config.rs:
```diff
@@ -0,0 +1,137 @@
+    let pub_key_path = path.join(&config.lms_pub_key);
+    gen_config.pub_keys.lms_pub_key = lms_pub_key_from_pem(&pub_key_path)?;
+
+    // Get the Private Keys.
```
This would be ported into an existing key signing environment. It will use the same mechanism there is, in terms of specifying the keys, for generating a firmware image.

nquarton replied:
Understood, thanks!

nquarton reviewed (commented):
Just 2 opens in previous comments (should dpe_result be added to AuthorizeAndStash response, can we add the command and response args to the RT readme?)

ArthurHeymans committed:
runtime: Document SET_AUTH_MANIFEST and AUTHORIZE_AND_STASH CMDs (#1631)

jhand2 reviewed (commented):
This is pretty hard to review as-is. Would you be able to split up this change as:

1. RT spec change
2. auth manifest tool
3. Runtime commands for using auth manifest

I think these are all pretty independent of each other

mhatrevi commented:
> This is pretty hard to review as-is. Would you be able to split up this change as:
> 
> 1. RT spec change
> 2. auth manifest tool
> 3. Runtime commands for using auth manifest
> 
> I think these are all pretty independent of each other

Not sure how that reduces the complexity of the review. These changes are related to each other and IMO, should be in a single PR. For ease of code review, I would suggest the following order of folders to review:

1. auth-manifest
2. runtime
3. Remaining folders

mhatrevi committed:
Merge branch 'main' into vmhatre/man-auth

mhatrevi committed:
Including manifest version in vendor endorsement calculation

mhatrevi committed:
Merge branch 'main' into vmhatre/man-auth

mhatrevi commented:
This PR will be broken down into the following two tasks:

1. Manifest tooling
2. Runtime changes for image authorization.

Closing this PR as such.
